clients/pkg/promtail/targets/kafka/target_syncer.go

```diff
@@ -0,0 +1,257 @@
+	}
+	topicManager, err := newTopicManager(client, cfg.KafkaConfig.Topics)
+	if err != nil {
+		return nil, fmt.Errorf("error creating topic manager: %w", err)
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+
+	t := &TargetSyncer{
+		logger:       logger,
+		ctx:          ctx,
+		cancel:       cancel,
+		topicManager: topicManager,
+		cfg:          cfg,
+		reg:          reg,
+		client:       pushClient,
+		close: func() error {
+			if err := group.Close(); err != nil {
+				level.Warn(logger).Log("msg", "error while closing consumer group", "err", err)
+			}
+			return client.Close()
+		},
+		consumer: consumer{
+			ctx:           context.Background(),
+			cancel:        func() {},
+			ConsumerGroup: group,
+			logger:        logger,
+		},
+	}
+	t.discoverer = t
+	t.loop()
+	return t, nil
+}
+
+func (ts *TargetSyncer) loop() {
+	topicChanged := make(chan []string)
+	ts.wg.Add(2)
+	go func() {
+		defer ts.wg.Done()
+		for {
+			select {
+			case <-ts.ctx.Done():
+				return
+			case topics := <-topicChanged:
+				level.Info(ts.logger).Log("msg", "new topics received", "topics", fmt.Sprintf("%+v", topics))
+				ts.stop()
+				if len(topics) > 0 { // no topics we don't need to start.
+					ts.start(ts.ctx, topics)
+				}
+			}
+		}
+	}()
+	go func() {
+		defer ts.wg.Done()
+		ticker := time.NewTicker(TopicPollInterval)
+		defer ticker.Stop()
+
+		tick := func() {
+			select {
+			case <-ts.ctx.Done():
+			case <-ticker.C:
+			}
+		}
+		for ; true; tick() { // instant tick.
+			if ts.ctx.Err() != nil {
+				ts.stop()
+				close(topicChanged)
+				return
+			}
+			newTopics, ok, err := ts.fetchTopics()
+			if err != nil {
+				level.Warn(ts.logger).Log("msg", "failed to fetch topics", "err", err)
+				continue
+			}
+			if ok {
+				topicChanged <- newTopics
+			}
+
+		}
+	}()
+}
+
+// fetchTopics fetches and return new topics, if there's a difference with previous found topics
+// it will return true as second return value.
+func (ts *TargetSyncer) fetchTopics() ([]string, bool, error) {
+	new, err := ts.topicManager.Topics()
+	if err != nil {
+		return nil, false, err
+	}
+	if len(ts.previousTopics) != len(new) {
+		ts.previousTopics = new
+		return new, true, nil
+	}
+	for i, v := range ts.previousTopics {
+		if v != new[i] {
+			ts.previousTopics = new
+			return new, true, nil
+		}
+	}
+	return nil, false, nil
+}
+
+func (ts *TargetSyncer) Stop() error {
+	ts.cancel()
+	ts.wg.Wait()
+	return ts.close()
+}
+
+// NewTarget creates a new targets based on the current kafka claim and group session.
+func (ts *TargetSyncer) NewTarget(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) (RunnableTarget, error) {
+	discoveredLabels := model.LabelSet{
+		"__topic":     model.LabelValue(claim.Topic()),
+		"__partition": model.LabelValue(fmt.Sprintf("%d", claim.Partition())),
+		"__member_id": model.LabelValue(session.MemberID()),
+		"__group_id":  model.LabelValue(ts.cfg.KafkaConfig.GroupID),
```

Comment:
Should these be further prefixed with `__meta_kafka`, like we have for `__meta_kubernetes_<etc>`?